File: content/snippets/typescript/design_patterns/memento.ts
class Memento {
    constructor(private state: string) {}
    
    getState(): string {
        return this.state;
    }
}

class Originator {
    private state: string = "";
    
    setState(state: string): void {
        this.state = state;
    }
    
    getState(): string {
        return this.state;
    }
    
    saveState(): Memento {
        return new Memento(this.state);
    }
    
    restoreState(memento: Memento): void {
        this.state = memento.getState();
    }
}

class Caretaker {
    private history: Memento[] = [];
    
    save(memento: Memento): void {
        this.history.push(memento);
    }
    
    restore(index: number): Memento | null {
        if (index >= 0 && index < this.history.length) {
            return this.history[index];
        }
        return null;
    }
}

const originator = new Originator();
const caretaker = new Caretaker();

originator.setState("State 1");
caretaker.save(originator.saveState());

originator.setState("State 2");
caretaker.save(originator.saveState());

originator.setState("State 3");
console.log(`Current: ${originator.getState()}`);

const restored = caretaker.restore(0);
if (restored) {
    originator.restoreState(restored);
    console.log(`Restored: ${originator.getState()}`);
}

File: content/snippets/cpp/design_patterns/memento.cpp
#include <iostream>
#include <string>
#include <vector>
#include <memory>

class Memento {
private:
    std::string state;
public:
    Memento(const std::string& s) : state(s) {}
    std::string getState() const { return state; }
};

class Originator {
private:
    std::string state;
public:
    void setState(const std::string& s) { state = s; }
    std::string getState() const { return state; }
    
    std::shared_ptr<Memento> saveState() {
        return std::make_shared<Memento>(state);
    }
    
    void restoreState(const std::shared_ptr<Memento>& memento) {
        state = memento->getState();
    }
};

class Caretaker {
private:
    std::vector<std::shared_ptr<Memento>> history;
public:
    void save(std::shared_ptr<Memento> memento) {
        history.push_back(memento);
    }
    
    std::shared_ptr<Memento> restore(int index) {
        if (index >= 0 && index < history.size()) {
            return history[index];
        }
        return nullptr;
    }
};

int main() {
    Originator originator;
    Caretaker caretaker;
    
    originator.setState("State 1");
    caretaker.save(originator.saveState());
    
    originator.setState("State 2");
    caretaker.save(originator.saveState());
    
    originator.setState("State 3");
    std::cout << "Current: " << originator.getState() << std::endl;
    
    originator.restoreState(caretaker.restore(0));
    std::cout << "Restored: " << originator.getState() << std::endl;
}
